Sisr demo

AvKhokhlov opened the pull request.

yozhijk committed:
merged light loading from remote master

AvKhokhlov committed:
Added super resolution post effect

AvKhokhlov committed:
Added base inference interface impl

AvKhokhlov committed:
fixed compile errors

AvKhokhlov committed:
renamed DenoiserType into PostProcessingType

AvKhokhlov committed:
remooved excessive code from application.cpp

AvKhokhlov committed:
sisr post effect passed into Baikal

AvKhokhlov committed:
fixed denoiser crush

AvKhokhlov committed:
fixies in super res

AvKhokhlov committed:
save intermidiate work

Commit:
added working resize and bug with not sisr mode

AvKhokhlov committed:
reworked interface for inference

AvKhokhlov committed:
made implementation for sisr model

AvKhokhlov committed:
Added new model interface (dont build)

AvKhokhlov committed:
added support of the 4th channel in super_res post effect

AvKhokhlov committed:
Added necessary libs, changes in inference logic

AvKhokhlov committed:
denoiser updatet with new inference and map/unmap ops for ml image buffers

AvKhokhlov committed:
Partial fix of some compilation errors, deleted setting of the input/output params for model

AvKhokhlov committed:
Fixed bug with CPU cache in super res and fixed incorrect image allocation in inderence

AvKhokhlov committed:
fixed render output size for denoiser and simple modes in standalone

AvKhokhlov committed:
yet another fix for buffer sizes in super res

AvKhokhlov committed:
almost working sisr

dtarakanov1 committed:
Fix input preparation

AvKhokhlov committed:
remooved 'git' word from denoise.cl

AvKhokhlov committed:
added bicubic

AvKhokhlov committed:
fix for frames from the past

AvKhokhlov committed:
fixed bug with too bright image after bicubic

AvKhokhlov committed:
fix for memory lick, Image struct now moveable

AvKhokhlov committed:
resize logic agregeated in separate function

AvKhokhlov committed:
split mode

AvKhokhlov committed:
Some refactoring (not compiled

AvKhokhlov committed:
save intermediate result

AvKhokhlov committed:
refactored denoise input

AvKhokhlov committed:
added sisr preproce class

AvKhokhlov committed:
renamed denoiser and super_res

AvKhokhlov committed:
added sisr_preproce implementation

AvKhokhlov committed:
refactoring in denoiser/sisr preproc, added new preproc logic into MlPostEffect

AvKhokhlov committed:
Moved image struct, ml_image replaced with Image in preprocess classes, added implementation in MlPostEffect entity

AvKhokhlov committed:
added MlPostEffect controll logic

AvKhokhlov committed:
made compile

AvKhokhlov committed:
added new logic with spp in post effect

AvKhokhlov committed:
added unsigned spp and stab for channels in sir mode

dtarakanov1 commented:
Look at the number of commits.

dtarakanov1 commented on 3rdparty/RadeonProML/RadeonProML.h:
```diff
@@ -0,0 +1,274 @@
+/**
+ * Context handle.
+ */
+typedef struct ml_context_t {}* ml_context;
```
Please take the latest file version.

dtarakanov1 commented on Baikal/CMakeLists.txt:
```diff
@@ -199,6 +203,7 @@ endif (BAIKAL_EMBED_KERNELS)
 
 target_compile_features(Baikal PRIVATE cxx_std_14)
 target_include_directories(Baikal PUBLIC "${Baikal_SOURCE_DIR}/Baikal")
+target_include_directories(Baikal PUBLIC "${Baikal_SOURCE_DIR}/Baikal")
```
A duplicate.

dtarakanov1 commented on Baikal/CMakeLists.txt:
```diff
@@ -233,5 +238,10 @@ else ()
 
+# add ml libs
+target_include_directories(Baikal PUBLIC "${Baikal_SOURCE_DIR}/3rdparty/RadeonProML")
+target_link_directories(Baikal PUBLIC "${Baikal_SOURCE_DIR}/3rdparty/RadeonProML")
```
It's too modern CMake command (v.3.13, and Baikal uses an older version). I'd propose
```
 target_link_libraries(Baikal PUBLIC "${Baikal_SOURCE_DIR}/3rdparty/libRadeonProML.so")
```

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                                               CLProgramManager const *program_manager,
+                                               std::uint32_t start_spp)
+        : DataPreprocess(context, program_manager, start_spp)
+        , m_primitives(CLWParallelPrimitives(context))
```
"`CLWParallelPrimitives`" is not required.

dtarakanov1 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -172,11 +194,74 @@ void CopyInterleaved(GLOBAL float4* restrict dst,
     }
+}
+
+static float3 ConvlutionCompute(float4 f0, float4 f1, float4 f2, float4 f3)
```
"ComputeConv **o** lution"

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+                                                            2 * src.GetElementCount());
+            }
+
+            auto scale_x = GetKernel("BicubicUpScaleX_x2");
```
Why can't we hide `BicubicUpScaleX/Y_x2` and expose a single kernel that calls both of them?

AvKhokhlov replied:
Because it's harder to implement and requires more weights in convolution

dtarakanov1 replied:
Is it impossible to call the two functions from another one?

AvKhokhlov replied:
Nope

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+
+            auto scale_x = GetKernel("BicubicUpScaleX_x2");
+
+            int argc = 0;
```
`unsigned argc = 0`. See the SetArg signature.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+        {
+            auto context = GetContext();
+
+            if (m_resizer_cache.GetElementCount() < 2 * src.GetElementCount())
```
What if the size is decreased?

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+                auto color = dynamic_cast<ClwOutput*>(input_set.at(OutputType::kColor))->data();
+                if (m_type == ModelType ::kDenoiser)
+                {
+                    context.CopyBuffer<float3>(0,
```
`<float3>` is not required.

dtarakanov1 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -149,8 +172,7 @@ void CopyInterleaved(GLOBAL float4* restrict dst,
                      int src_channels_num,
-                     int channels_to_copy,
-                     GLOBAL float* restrict out_sample_count)
+                     int channels_to_copy)
```
I wonder why all of these values are `int`s. None of them has negative value semantics.
(Do not change so far)

dtarakanov1 commented on Baikal/PostEffects/ML/data_preprocess.cpp:
```diff
@@ -0,0 +1,76 @@
+        {
+            auto copy_kernel = GetKernel("CopyInterleaved");
+
+            int argc = 0;
```
`unsigned`

dtarakanov1 commented on Baikal/PostEffects/ML/data_preprocess.cpp:
```diff
@@ -0,0 +1,76 @@
+                                         copy_kernel);
+        }
+    }
+}
```
No newline in the end.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+        using OutputType = Renderer::OutputType;
+
+        DenoiserPreprocess::DenoiserPreprocess(CLWContext context,
+                                               CLProgramManager const *program_manager,
```
Move `*` to the left.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+            float real_sample_count = 0.f;
+            bool too_few_samples = false;
+
+            if (!m_is_init)
```
It's not clear what the `m_is_init` means without looking at the code that uses it (is initial state / is initializing / is initialized?).

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+
+                        if (real_sample_count < m_start_spp)
+                        {
+                            too_few_samples = true;
```
Why can't we immediately return here instead setting the flag?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                    case OutputType::kColor:
+                    {
+                        DivideBySampleCount(CLWBuffer<float3>::CreateFromClBuffer(m_cache),
+                                            CLWBuffer<float3>::CreateFromClBuffer(device_mem));
```
Isn't device_mem the float3 already?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                                      m_height,
+                                      channels_count,
+                                      m_channels,
+                                      0,
```
Could you please add comments with the parameter names?
```
...
m_input /*param0*/,
...
0 /*param1*/,
4 /*param2*/,
...
```

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                                      m_channels,
+                                      0,
+                                      4,
+                                      1).Wait();
```
Why do we wait here?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                        break;
+                    }
+                    case OutputType::kViewShadingNormal:
+                    {
```
These and subsequent curly braces are not required. I'd also remove the braces above after eliminating the temporary variables.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                                      4,
+                                      1).Wait();
+
+                        channels_count += 1;
```
Why `+=`?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                    }
+                    case OutputType::kViewShadingNormal:
+                    {
+                        channels_to_copy = 2;
```
Generally this is always `desc.second`.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+                throw std::runtime_error("map operation failed");
+            }
+
+            context.ReadBuffer<float>(0,
```
`<float>` is not required.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+            switch (m_model)
+            {
+                case Model::kColorDepthNormalGloss7:
+                    return std::set<Renderer::OutputType>(
```
DRY. These data can be taken from the `m_layout`.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+        void DenoiserPreprocess::DivideBySampleCount(CLWBuffer<float3> dst,
+                                                     CLWBuffer<float3> src)
+        {
+            assert (dst.GetElementCount() >= src.GetElementCount());
```
`assert` is not an operator in C++, so the space is before the bracket is extra.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,291 @@
+            auto division_kernel = GetKernel("DivideBySampleCount");
+
+            // Set kernel parameters
+            int argc = 0;
```
`unsigned`

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Wrong copyright indentation.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+
+            using MemoryLayout = std::vector<std::pair<Renderer::OutputType, std::size_t>>;
+
+            void DivideBySampleCount(CLWBuffer<RadeonRays::float3> dst,
```
`const` reference.

AvKhokhlov replied:
It is referenced object, and also the function changes its value

dtarakanov1 replied:
If the buffer was a pointer, e.g. `char*`, adding `const` this way would result in `char* const`, not in `const char*` (and so it would be OK). And you do not change the buffer value here, you change it contents, so the analogy is fully valid.

Furthermore if you assume value semantics for the CLW buffer, you create a new buffer here and change it, so the original buffer is untouched (which is totally wrong).

Last but not least, copying a CLW buffer results in extra calls to clRetainBuffer/clReleaseBuffer functions (which e.g. is heavier than atomic increment/decrement for std::shared_ptr).

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+            using MemoryLayout = std::vector<std::pair<Renderer::OutputType, std::size_t>>;
+
+            void DivideBySampleCount(CLWBuffer<RadeonRays::float3> dst,
+                                     CLWBuffer<RadeonRays::float3> src);
```
`const` reference.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+
+#include "CLW.h"
+#include "Utils/clw_class.h"
+#include "data_preprocess.h"
```
Please use a longer path version.

AvKhokhlov replied:
?

dtarakanov1 replied:
`PostEffects/ML/denoiser_preprocess.h`

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+
+#pragma once
+
+#include "CLW.h"
```
Move right before the std headers (because it's an external library header).

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,81 @@
+            MemoryLayout m_layout;
+            CLWBuffer<float> m_cache;
+            CLWBuffer<float> m_input;
+            ml_context m_context;
```
I propose to write small RAII wrappers for ML handles. Approximately this way:
```
template<class T>
using Handle = std::unique_ptr<typename std::remove_pointer<T>::type, void (*)(T)>;
...
Handle<ml_context> m_context(mlCreateContext(), mlReleaseContext);
```

dtarakanov1 commented on Baikal/PostEffects/ML/image.cpp:
```diff
@@ -0,0 +1,59 @@
+ THE SOFTWARE.
+ ********************************************************************/
+
+#include "image.h"
```
A path relative to the library root?

dtarakanov1 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -0,0 +1,168 @@
+            m_input_desc.dtype = image_info.dtype;
+            if (mlSetModelInputInfo(m_model.GetModel(), &m_input_desc) != ML_OK)
+            {
+                throw std::runtime_error(
```
Use `CheckModelStatus()` everywhere.

dtarakanov1 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -0,0 +1,168 @@
+
+void CheckModelStatus(ml_model model, bool status)
+{
+    if (!status)
```
Since the success status is always `ML_OK`, it would be less verbose (in the client code) to pass the ML status here and compare it with `ML_OK`.

dtarakanov1 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -0,0 +1,168 @@
+
+namespace {
+
+void CheckModelStatus(ml_model model, bool status)
```
This can be made a private member.

dtarakanov1 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -0,0 +1,168 @@
+                {
+                    const size_t size = 1024;
+                    char buffer[size];
+                    mlGetModelError(m_model.GetModel(), buffer, size);
```
This code repeats the CheckModelStatus.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+ THE SOFTWARE.
+ ********************************************************************/
+
+#include "ml_post_effect.h"
```
Longer paths?

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+    /**********************************************************************
```
Oops.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+    /**********************************************************************
+ Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
2019 or 2018-2019

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+        , m_last_seq(0)
+        , m_program(program_manager)
+        {
+            m_inference = nullptr;
```
Why not within the initializer list?

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+            {
+                case ModelType ::kDenoiser:
+                    m_preproc = std::unique_ptr<DataPreprocess>(
+                            new DenoiserPreprocess(GetContext(), m_program));
```
Use `std::make_unique` here as well.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+            // init preprocessing
+            switch (m_type)
+            {
+                case ModelType ::kDenoiser:
```
Extra space here and below.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+            switch (m_type)
+            {
+                case ModelType::kDenoiser:
+                    return std::unique_ptr<Inference>(
```
`std::make_unique`?

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+            if (res.image != ML_INVALID_HANDLE && res.tag >= m_start_seq)
+            {
+                size_t res_size;
+                auto res_data = static_cast<float *>(mlMapImage(res.image, &res_size));
```
`<float*>`

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+
+                auto dest = m_host.data();
+                auto source = res_data;
+                for (auto i = 0u; i < shape.width * shape.height; ++i)
```
`FillBuffer` + `CopyInterleaved`

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+
+                mlUnmapImage(res.image, res_data);
+
+                context.WriteBuffer<float3>(0,
```
`<float3>` is not required.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+
+            if (m_has_denoised_img)
+            {
+                context.CopyBuffer<float3>(0,
```
`<float3>` is not required.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.h:
```diff
@@ -0,0 +1,70 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Bad indentation.
2019

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.h:
```diff
@@ -0,0 +1,70 @@
+        enum class ModelType
+        {
+            kDenoiser,
+            kSisr
```
`kSisr,`

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,98 @@
-#include "PostEffects/ML/model_holder.h"
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation. 2019.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,98 @@
+                                 std::string const& visible_devices)
         {
-            if (g_load_model != nullptr)
+            try
```
`try` may be moved one line up, and `catch` one line down, this will decrease the amount of indentation.
Actually `try... catch` is not required at all if the `m_context` and `m_model` are RAII-wrappers (there is a corresponding comment)`.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,98 @@
+
+                m_model = mlCreateModel(m_context, &params);
+
+                if (m_model == ML_INVALID_HANDLE)
```
The error message may be taken from the context (`mlGetContextError`). And same below.
So there could be `CheckContextStatus(ml_status)`.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,98 @@
             {
-                throw std::runtime_error(std::string("Library error: ") + dlerror());
+                ShutDown();
+                throw ex;
```
Just `throw;`

dtarakanov1 commented on Baikal/RenderFactory/render_factory.h:
```diff
@@ -36,7 +36,8 @@ namespace Baikal
         kWaveletDenoiser,
-        kMLDenoiser
+        kMLDenoiser,
+        kSISR
```
`,`

dtarakanov1 commented on Baikal/Utils/clw_class.h:
```diff
@@ -36,6 +36,7 @@ namespace Baikal
         virtual ~ClwClass() = default;
 
         CLWContext GetContext() const { return m_context; }
+        const CLProgramManager* GetProgram() const { return m_program_manager; }
```
Where is this used?

dtarakanov1 commented on BaikalStandalone/Application/app_utils.cpp:
```diff
@@ -189,21 +191,25 @@ namespace Baikal
+                s.post_processing_type = PostProcessingType::kWavelet;
             }
-            else if (denoiser_type == "wavelet")
+            else if (post_processing_type == "ml")
```
ML is not a postprocessing.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -10,34 +32,35 @@ namespace Baikal
-                       std::string const& visible_devices);
 
-            ML::Model* operator ->() const
+            ml_model GetModel()
```
`const`?

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -10,34 +32,35 @@ namespace Baikal
-            {
-                return *m_model;
-            }
+            ml_image CreateImage(ml_image_info const& info);
```
`const`?

dtarakanov1 replied:
Up.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -10,34 +32,35 @@ namespace Baikal
+            ModelHolder& operator = (const ModelHolder&) = delete;
+            ModelHolder& operator = (ModelHolder&&) = delete;
+
+            ml_context m_context;
```
Isn't private?

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,149 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation. 2019.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,149 @@
+
+            // read spp from first pixel as 4th channel
+            RadeonRays::float3 pixel = .0f;
+            context.ReadBuffer<float3>(0, clw_input->data(), &pixel, 1).Wait();
```
`<float3>` is not required.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,149 @@
+                          CLWBuffer<float>::CreateFromClBuffer(m_cache),
+                          m_width,
+                          m_height,
+                          0,  // dst channels offset
```
I recommend `0 /*exact_parameter_name*/,`
- Exact name to make it easier to search.
- /* */ comment before the comma to avoid bad formatting by some IDEs.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,149 @@
+                 throw std::runtime_error("map operation failed");
+             }
+
+             context.ReadBuffer<float>(0,
```
`<float>` is excessive.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,149 @@
+            auto tonemapping = GetKernel("TonemapExponential");
+
+            // Set kernel parameters
+            int argc = 0;
```
`unsigned`

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.h:
```diff
@@ -0,0 +1,63 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.h:
```diff
@@ -0,0 +1,63 @@
+        private:
+            void Init(std::uint32_t width, std::uint32_t height);
+
+            void Tonemap(CLWBuffer<RadeonRays::float3> dst,
```
Pass as const refs.

dtarakanov1 commented on BaikalStandalone/Application/app_utils.h:
```diff
@@ -90,6 +90,9 @@ namespace Baikal
         std::string base_image_file_name;
         std::string image_file_format;
 
+        //light file
```
Too obvious.

dtarakanov1 commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -82,8 +85,11 @@ namespace Baikal
-        m_width = (std::uint32_t)settings.width;
-        m_height = (std::uint32_t)settings.height;
+        m_width = (m_post_processing_type == PostProcessingType::kSISR) ?\
+            (std::uint32_t)settings.width / 2 : (std::uint32_t)settings.width;
```
It's better to make settings.width/heigth uint32_t.

dtarakanov1 commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -154,15 +160,22 @@ namespace Baikal
+        if (m_post_processing_type != PostProcessingType::kSISR)
+        {
+            m_post_effect_output = m_cfgs[device_idx].factory->CreateOutput(m_width, m_height);
+        } else
```
...

dtarakanov1 commented on BaikalStandalone/Utils/config_manager.h:
```diff
@@ -41,12 +41,13 @@ enum class DeviceType
     kWavelet,
-    kML
+    kML,
+    kSISR
```
`,`

AvKhokhlov committed:
working sisr

AvKhokhlov committed:
working denoiser

AvKhokhlov committed:
fixed bug with spp delay

AvKhokhlov committed:
fixed black screen in the left part in split sisr mode

AvKhokhlov committed:
code review (part 1)

AvKhokhlov committed:
code review (part 2)

AvKhokhlov committed:
code review (part 3)

AvKhokhlov committed:
code review (part 4)

dtarakanov1 commented on Baikal/CMakeLists.txt:
```diff
@@ -233,5 +237,10 @@ else ()
+# add ml libs
+target_include_directories(Baikal PUBLIC "${Baikal_SOURCE_DIR}/3rdparty/RadeonProML")
+target_link_libraries(Baikal PUBLIC "${Baikal_SOURCE_DIR}/3rdparty/RadeonProML/libRadeonProML.so")
+target_link_libraries(Baikal PUBLIC libRadeonProML.so)
```
This is not required anymore.

dtarakanov1 commented on Baikal/PostEffects/ML/data_preprocess.cpp:
```diff
@@ -0,0 +1,76 @@
+        {  }
+
+
+        CLWEvent DataPreprocess::WriteToInputs(CLWBuffer<float> dst_buffer,
```
Pass buffers by const reference.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+/**********************************************************************
+ Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
2019

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+            ml_image_info image_info = {ML_FLOAT32, m_width, m_height, m_channels};
+            m_image = mlCreateImage(m_context.get(), &image_info);
+
+            if (!m_image)
```
CheckContextStatus?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+                                      m_cache,
+                                      m_width,
+                                      m_height,
+                                      channels_count, // dst channels offset
```
Some IDEs break formatting. So it's better to write
```
4 /*src_channels_num*/,
```

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+                    }
+                    case OutputType::kDepth:
+                    {
+                        auto normalized_buf = CLWBuffer<cl_float3>::CreateFromClBuffer(m_cache);
```
This variable may be eliminated: it's used actually once in Normalize, the 2nd usage is a "cast", it can be done from m_cache directly.
When the variable is eliminated, the surrounding curly brackets can be removed, as well as brackets for other `case`s.

dtarakanov1 replied:
This variable may be eliminated: it is used once in Normalize, the 2nd usage is a "cast", and the "cast" may be done directly from m_cache. But actually the m_cache is float already :)

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+                    case OutputType::kViewShadingNormal:
+                    case OutputType::kGloss:
+                    case OutputType::kAlbedo:
+                    {
```
The curly brackets may be removed since they are not needed.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+                                            CLWBuffer<float3>::CreateFromClBuffer(device_mem));
+
+                        WriteToInputs(m_input,
+                                      CLWBuffer<float>::CreateFromClBuffer(m_cache),
```
Already CLWBuffer\<float>?

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+        Image DenoiserPreprocess::MakeInput(PostEffect::InputSet const& inputs)
+        {
+            auto context = GetContext();
+            unsigned channels_count = 0u;
```
It's always used as "channel_offset", so might be renamed.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+
+        std::tuple<std::uint32_t, std::uint32_t> DenoiserPreprocess::ChannelsNum() const
+        {
+            return std::tuple<std::uint32_t, std::uint32_t>(m_channels, 3);
```
```C++
return {m_channels, 3};
```

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,82 @@
+/**********************************************************************
+Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation.

dtarakanov1 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,82 @@
+            CLWBuffer<float> m_cache;
+            CLWBuffer<float> m_input;
+            Handle<ml_context> m_context;
+            ml_image m_image;
```
Why not Handle?

dtarakanov1 commented on Baikal/PostEffects/ML/image.cpp:
```diff
@@ -0,0 +1,51 @@
+        : tag(tag), image(image)
+        {}
+
+        Image::Image(Image &&img)
```
If `image` is a handle, all the moving code is not required (unless tag = 0 is mandatory for an "moved-away" state).

dtarakanov1 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -0,0 +1,149 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indent, 2019.

dtarakanov1 commented on Baikal/PostEffects/ML/inference.h:
```diff
@@ -1,9 +1,36 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indent, 2019.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+                auto dest = m_host.data();
+                auto source = res_data;
+                auto output_shape = m_inference->GetOutputShape();
+                for (auto i = 0u; i < output_shape.width * output_shape.height; ++i)
```
CopyInterleaved

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+
+                if (m_type == ModelType ::kDenoiser)
+                {
+                    context.CopyBuffer<float3>(0,
```
`<float3>` is not needed.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+        }
+        void MlPostEffect::Resize_x2(CLWBuffer<RadeonRays::float3> dst, CLWBuffer<RadeonRays::float3> src)
+        {
+            auto context = GetContext();
```
I'd eliminate this variable.

dtarakanov1 commented on Baikal/PostEffects/ML/ml_post_effect.h:
```diff
@@ -0,0 +1,70 @@
+/**********************************************************************
+Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,73 @@
-#include "PostEffects/ML/model_holder.h"
+/**********************************************************************
+Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved.
```
i

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,73 @@
+
+        ml_image ModelHolder::CreateImage(ml_image_info const& info)
+        {
+            auto tensor = mlCreateImage(m_context.get(), &info);
```
`image`

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.cpp:
```diff
@@ -1,99 +1,73 @@
-            params.visible_devices = visible_devices.c_str();
-            m_model.reset(g_load_model(&params));
-        }
+        ModelHolder::~ModelHolder()
```
Not needed?

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -1,6 +1,28 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
Indentation, 2019.

dtarakanov1 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -10,34 +32,36 @@ namespace Baikal
-                       std::string const& visible_devices);
 
-            ML::Model* operator ->() const
+            const ml_model GetModel()
```
`const` is excessive.
But the method might be `const`.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+/**********************************************************************
+Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved.
```
i

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+            ml_image_info image_info = {ML_FLOAT32, width, height, 3};
+            m_image = mlCreateImage(m_context.get(), &image_info);
+
+            if (!m_image)
```
CheckContextStatus

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+            auto context = GetContext();
+
+            // read spp from first pixel as 4th channel
+            RadeonRays::float3 pixel = .0f;
```
Mya be extracted to `unsigned ReadSampleCount(...)`.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+
+            // delete 4th channel
+            WriteToInputs(m_input,
+                          CLWBuffer<float>::CreateFromClBuffer(m_cache),
```
Isn't a float buffer already.

dtarakanov1 commented on Baikal/PostEffects/ML/sisr_preprocess.h:
```diff
@@ -0,0 +1,65 @@
+/**********************************************************************
+Copyright (c) 2019 Advanced Micro Devices, Inc. All rights reserved.
```
i

dtarakanov1 commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -770,36 +755,39 @@ namespace Baikal
             AddPostEffect(m_primary, PostEffectType::kWaveletDenoiser);
             break;
-        case DenoiserType::kML:
+        case PostProcessingType::kML:
```
`kML` is inconsistent, because SISR is also ML.

achernigin1987 replied:
I propose kBilateralDenoiser, kWaveletDenoser, kMLDenoiser and kMLUpsample

dtarakanov1 commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -175,72 +189,29 @@ namespace Baikal
-                settings.camera_pos
-                , settings.camera_at
-                , settings.camera_up);
+            case CameraType::kPerspective:
```
Please revert the indent.

dtarakanov1 commented on BaikalStandalone/Application/cl_render.cpp:
```diff
@@ -770,36 +755,39 @@ namespace Baikal
             m_post_effect->SetParameter("start_spp", settings.denoiser_start_spp);
             break;
+        case PostProcessingType::kSISR:
+            AddPostEffect(m_primary, PostEffectType::kSISR);
```
Set `start_spp`?

dtarakanov1 commented on BaikalStandalone/Application/scene_load_utils.cpp:
```diff
@@ -0,0 +1,148 @@
+/**********************************************************************
+ Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
2019

dtarakanov1 commented on BaikalStandalone/Application/scene_load_utils.h:
```diff
@@ -0,0 +1,34 @@
+
```
Extra line.

dtarakanov1 commented on BaikalStandalone/Application/scene_load_utils.cpp:
```diff
@@ -0,0 +1,148 @@
+    LoadMaterials(basepath, scene);
+    LoadLights(settings.light_file, scene);
+    return scene;
+}
```
No EOL.

dtarakanov1 commented on BaikalStandalone/Application/scene_load_utils.h:
```diff
@@ -0,0 +1,34 @@
+
+/**********************************************************************
+ Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
```
2019

achernigin1987 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -172,11 +194,74 @@ void CopyInterleaved(GLOBAL float4* restrict dst,
     }
+}
+
+static float3 ConvolutionCompute(float4 f0, float4 f1, float4 f2, float4 f3)
```
ConvolutionCompute -> BicubicConvolutionCompute

achernigin1987 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -172,11 +194,74 @@ void CopyInterleaved(GLOBAL float4* restrict dst,
+}
+
+KERNEL
+void BicubicUpScaleX_x2(// size of the dst buffer should be enough
```
BicubicUpScaleX_x2 -> BicubicUpscale2x_X

achernigin1987 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -172,11 +194,74 @@ void CopyInterleaved(GLOBAL float4* restrict dst,
+
+
+KERNEL
+void BicubicUpScaleY_x2(// size of the dst buffer should be enough to store
```
BicubicUpScaleY_x2 -> BicubicUpscale2x_Y

achernigin1987 commented on Baikal/PostEffects/ML/data_preprocess.h:
```diff
@@ -0,0 +1,73 @@
+{
+    namespace PostEffects
+    {
+        class DataPreprocess : public ClwClass
```
DataPreprocess -> DataPreprocessor or Preprocessor. Verbs is preferable for methods

achernigin1987 commented on Baikal/PostEffects/ML/data_preprocess.h:
```diff
@@ -0,0 +1,73 @@
+            // returns channels num at input and output
+            virtual std::tuple<std::uint32_t, std::uint32_t> ChannelsNum() const = 0;
+
+            void ResetSpp(std::uint32_t spp)
```
ResetSpp -> SetStartSpp,  spp -> start_spp

achernigin1987 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,82 @@
+            kColorAlbedoDepthNormal9
+        };
+
+        class DenoiserPreprocess: public DataPreprocess
```
DenoiserPreprocess -> DenoiserPreprocessor

achernigin1987 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+#include "Output/clwoutput.h"
+#include "math/mathutils.h"
+
+#include <RadeonProML.h>
```
why "<...>" Is it system header ?

achernigin1987 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+                m_width = color->width();
+                m_height = color->height();
+                Init(m_width, m_height);
+                m_is_initialized = true;
```
Why do you set this flag outside "Init" function ?

dtarakanov1 replied:
That was my comment from several months ago.
This way you can see why `Init()` is not called every time. You cannot see right here if `Init()` sets the flag.

achernigin1987 commented on Baikal/PostEffects/ML/denoiser_preprocess.h:
```diff
@@ -0,0 +1,82 @@
+        private:
+            void Init(std::uint32_t width, std::uint32_t height);
+
+            using MemoryLayout = std::vector<std::pair<Renderer::OutputType, std::size_t>>;
```
Please add comment that size_t is channels num

dtarakanov1 replied:
...and make it `int` :)

achernigin1987 commented on Baikal/PostEffects/ML/denoiser_preprocess.cpp:
```diff
@@ -0,0 +1,252 @@
+
+            if (!m_image)
+            {
+                throw std::runtime_error("can not create ml_image");
```
Couldn't...

achernigin1987 commented on Baikal/PostEffects/ML/image.h:
```diff
@@ -0,0 +1,45 @@
+            Image();
+            Image(std::uint32_t tag, ml_image image);
+            Image(Image &&);
+            //~Image();
```
Is it needed ?

achernigin1987 commented on Baikal/PostEffects/ML/inference.h:
```diff
@@ -15,12 +42,40 @@ namespace Baikal
+
+            Image PopOutput();
+            virtual ~Inference();
+
```
empty line

achernigin1987 commented on Baikal/PostEffects/ML/ml_post_effect.h:
```diff
@@ -0,0 +1,70 @@
+            kSisr,
+        };
+
+        class MlPostEffect : public ClwPostEffect
```
M**L**PostEffect

achernigin1987 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -0,0 +1,272 @@
+        {
+            return m_preproc->GetInputTypes();
+        }
+        void MlPostEffect::Resize_x2(CLWBuffer<RadeonRays::float3> dst, CLWBuffer<RadeonRays::float3> src)
```
missing empty line

achernigin1987 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -1,6 +1,28 @@
 #pragma once
 
-#include "PostEffects/ML/model.h"
+#include <RadeonProML.h>
```
Why "<...>" ?

achernigin1987 commented on Baikal/PostEffects/ML/model_holder.h:
```diff
@@ -10,34 +32,36 @@ namespace Baikal
+
+            ~ModelHolder();
+
+
```
empty line

achernigin1987 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+            return Image(sample_count, m_image);
+        }
+
+        void SisrPreprocess::Tonemap(CLWBuffer<RadeonRays::float3> dst,
```
There is no such word "tonemap". DoToneMapping or ApplyToneMapping ?

achernigin1987 commented on Baikal/PostEffects/ML/sisr_preprocess.cpp:
```diff
@@ -0,0 +1,156 @@
+        {
+            assert (dst.GetElementCount() >= src.GetElementCount());
+
+            auto tonemapping = GetKernel("TonemapExponential");
```
There is no such word "Tonemap"

achernigin1987 commented on Baikal/PostEffects/ML/ml_post_effect.h:
```diff
@@ -0,0 +1,70 @@
+        enum class ModelType
+        {
+            kDenoiser,
+            kSisr,
```
kSisr -> kUpsampler

achernigin1987 commented on BaikalStandalone/Application/app_utils.cpp:
```diff
@@ -189,21 +191,25 @@ namespace Baikal
+                s.post_processing_type = PostProcessingType::kML;
             }
-            else if (denoiser_type == "ml")
+            else if (post_processing_type == "sisr")
```
"sisr" -> "upsample"

AvKhokhlov committed:
Code review (part 5)

AvKhokhlov committed:
code review (part 6)

AvKhokhlov committed:
updated license

achernigin1987 commented on Baikal/PostEffects/ML/sisr_preprocessor.cpp:
```diff
@@ -31,14 +31,14 @@ namespace Baikal
         using float3 =  RadeonRays::float3;
 
-        SisrPreprocess::SisrPreprocess(CLWContext context,
+        SisrPreprocessor::SisrPreprocessor(CLWContext context,
```
UpsamplerPreprocessor

achernigin1987 commented on Baikal/PostEffects/ML/sisr_preprocessor.cpp:
```diff
@@ -125,12 +125,12 @@ namespace Baikal
             assert (dst.GetElementCount() >= src.GetElementCount());
 
-            auto tonemapping = GetKernel("TonemapExponential");
+            auto tonemapping = GetKernel("ToneMapingExponential");
```
ToneMap**p**ingExponential

achernigin1987 commented on Baikal/PostEffects/ML/ml_post_effect.cpp:
```diff
@@ -216,18 +216,19 @@ namespace Baikal
         }
-        void MlPostEffect::Resize_x2(CLWBuffer<RadeonRays::float3> dst, CLWBuffer<RadeonRays::float3> src)
+
+        void MLPostEffect::Resize_x2(CLWBuffer<RadeonRays::float3> dst, CLWBuffer<RadeonRays::float3> src)
```
Resize_**2x**

achernigin1987 commented on Baikal/Kernels/CL/denoise.cl:
```diff
@@ -114,7 +114,7 @@ void BilateralDenoise_main(
 
 KERNEL
-void TonemapExponential(GLOBAL float4* restrict dst,
+void ToneMapingExponential(GLOBAL float4* restrict dst,
```
ToneMap**p**ingExponential

achernigin1987 commented on Baikal/PostEffects/ML/inference.cpp:
```diff
@@ -23,7 +23,7 @@ THE SOFTWARE.
 #include "PostEffects/ML/error_handler.h"
 #include "inference.h"
 #include <cassert>
```
<casssert>  should be after other our headers

AvKhokhlov committed:
Code review (part 7)

AvKhokhlov commented:
- Added upsampler mode using neural network
- Added RadeonProML lib for working with ML models
- Reworked architecture of the ML part
- Significant refactoring of the ML part
